File: desktop/macos-app/Sources/Features/Login/SSOWebView.swift
// Gerege Template Platform V3.0
// Gerege Systems Development Team болон Claude AI хамтран бүтээв, 2026.

// Gerege SSO нэвтрэлт — вэбтэй ЯГ ИЖИЛ урсгал, бүхэлдээ платформын BFF-ээр:
//
//   /api/auth/sso/start → Gerege SSO (нэвтрэлт) → /sso/callback (cookie суулгана)
//   → /me/dashboard
//
// Апп нь SSO дээр ӨӨРИЙН OIDC client бүртгүүлэхгүй (native/PKCE урсгал байхгүй) —
// вэб client-ийн логикийг тэр чигт нь ашиглана. Ингэснээр нэвтрэлтийн бодлого
// нэг л газар (BFF) байрлана.
//
// Дашбоард руу шилжих АГШИНД навигацыг зогсоож, WKWebView-ийн cookie-г
// `HTTPCookieStorage.shared` руу хуулна: цаашид APIClient тэр session-оор
// ажиллана. Вэб дашбоардыг апп дотор РЕНДЭРЛЭХГҮЙ — native дэлгэц рүү шилжинэ.
//
// Энэ нь `mobile/ios/TemplateApp/Sources/SSOWebView.swift`-ийн macOS хувилбар:
// UIViewRepresentable → NSViewRepresentable, бусад логик ижил.

import SwiftUI
import WebKit

struct SSOWebView: NSViewRepresentable {
    let onDone: () -> Void
    let onCancel: () -> Void

    func makeCoordinator() -> Coordinator {
        Coordinator(onDone: onDone, onCancel: onCancel)
    }

    func makeNSView(context: Context) -> WKWebView {
        let config = WKWebViewConfiguration()
        // Нэвтрэлтийн cookie нь тогтвортой сан руу суух ёстой — эс бөгөөс апп
        // хаагдахад session алдагдана.
        config.websiteDataStore = .default()
        let web = WKWebView(frame: .zero, configuration: config)
        web.navigationDelegate = context.coordinator
        web.load(URLRequest(url: APIClient.baseURL.appendingPathComponent("/api/auth/sso/start")))
        return web
    }

    func updateNSView(_ nsView: WKWebView, context: Context) {}

    final class Coordinator: NSObject, WKNavigationDelegate {
        private let onDone: () -> Void
        private let onCancel: () -> Void
        private var finished = false

        init(onDone: @escaping () -> Void, onCancel: @escaping () -> Void) {
            self.onDone = onDone
            self.onCancel = onCancel
        }

        func webView(_ webView: WKWebView,
                     decidePolicyFor navigationAction: WKNavigationAction,
                     decisionHandler: @escaping (WKNavigationActionPolicy) -> Void) {
            guard let url = navigationAction.request.url else {
                decisionHandler(.allow)
                return
            }
            // Cookie суусны дараа BFF нь /me/... руу шилжүүлнэ — яг энд зогсооно.
            if url.host == APIClient.baseURL.host, url.path.hasPrefix("/me"), !finished {
                finished = true
                decisionHandler(.cancel)
                Task { @MainActor in
                    await Self.syncCookies(from: webView)
                    onDone()
                }
                return
            }
            decisionHandler(.allow)
        }

        func webView(_ webView: WKWebView, didFail navigation: WKNavigation!, withError error: Error) {
            guard !finished else { return }
            finished = true
            Task { @MainActor in onCancel() }
        }

        func webView(_ webView: WKWebView, didFailProvisionalNavigation navigation: WKNavigation!, withError error: Error) {
            guard !finished else { return }
            // Навигацыг өөрсдөө цуцалсан үед (NSURLErrorCancelled) алдаа гэж үзэхгүй.
            if (error as NSError).code == NSURLErrorCancelled { return }
            finished = true
            Task { @MainActor in onCancel() }
        }

        /// WKWebView-ийн cookie сан → HTTPCookieStorage.shared (URLSession уншина).
        @MainActor
        private static func syncCookies(from webView: WKWebView) async {
            let store = webView.configuration.websiteDataStore.httpCookieStore
            let cookies: [HTTPCookie] = await withCheckedContinuation { cont in
                store.getAllCookies { cont.resume(returning: $0) }
            }
            for cookie in cookies { HTTPCookieStorage.shared.setCookie(cookie) }
        }
    }
}

File: mobile/android/TemplateApp/app/src/main/java/mn/gerege/temp/SsoWebView.kt
// Gerege Template Platform V3.0
// Gerege Systems Development Team болон Claude AI хамтран бүтээв, 2026.

package mn.gerege.temp

import android.annotation.SuppressLint
import android.graphics.Bitmap
import android.webkit.CookieManager
import android.webkit.WebResourceRequest
import android.webkit.WebView
import android.webkit.WebViewClient
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.viewinterop.AndroidView

// Gerege SSO нэвтрэлт — вэбтэй ЯГ ИЖИЛ урсгал, бүхэлдээ template BFF-ээр:
//   /api/auth/sso/start → Gerege SSO (нэвтрэлт) → /sso/callback (cookie суулгана)
//   → /me/dashboard.
// Апп нь SSO дээр өөрийн OIDC client БҮРТГҮҮЛЭХГҮЙ (native/PKCE урсгал байхгүй) —
// template-ийн вэб client-ийн логикийг тэр чигт нь ашиглана. Дашбоард руу шилжих
// агшинд навигацыг зогсоож, cookie-г диск рүү flush хийгээд native дэлгэц рүү
// шилжинэ (вэб дашбоардыг апп дотор рендэрлэхгүй).
@SuppressLint("SetJavaScriptEnabled") // SSO хуудас JS шаарддаг; зөвхөн өөрсдийн BFF/SSO домэйн ачаална
@Composable
fun SsoWebView(modifier: Modifier = Modifier, onDone: () -> Unit) {
    AndroidView(
        modifier = modifier,
        factory = { ctx ->
            val web = WebView(ctx)
            web.settings.javaScriptEnabled = true
            web.settings.domStorageEnabled = true
            // JS гүүр (addJavascriptInterface) ОГТ нэмэхгүй — вэб хуудас native
            // код руу хүрэх гарц байх ёсгүй.
            CookieManager.getInstance().setAcceptCookie(true)
            CookieManager.getInstance().setAcceptThirdPartyCookies(web, true)
            web.webViewClient = SsoClient(onDone)
            web.loadUrl(ApiClient.ssoStartUrl)
            web
        },
    )
}

private class SsoClient(private val onDone: () -> Unit) : WebViewClient() {
    private var finished = false

    override fun shouldOverrideUrlLoading(view: WebView, request: WebResourceRequest): Boolean =
        complete(request.url.toString())

    // Сервер талын redirect зарим тохиолдолд shouldOverrideUrlLoading-д ирэхгүй
    // байж болзошгүй тул хуудас эхлэх агшинд ч шалгана (хамгаалалтын тор).
    override fun onPageStarted(view: WebView, url: String, favicon: Bitmap?) {
        if (complete(url)) view.stopLoading() else super.onPageStarted(view, url, favicon)
    }

    /** Нэвтрэлт дуусах хаяг мөн бол cookie-г flush хийж WebView-г хаана. */
    private fun complete(url: String): Boolean {
        if (finished || !SsoPolicy.isSignedInUrl(ApiClient.baseUrl, url)) return false
        finished = true
        CookieManager.getInstance().flush()
        onDone()
        return true
    }
}
